File: k2/python/host/csrc/fsa_algo.cc
// k2/python/host/csrc/fsa_algo.cc

// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

// See ../../../LICENSE for clarification regarding multiple authors

#include "k2/python/host/csrc/fsa_algo.h"

#include <memory>
#include <utility>

#include "k2/csrc/host/arcsort.h"
#include "k2/csrc/host/array.h"
#include "k2/csrc/host/connect.h"
#include "k2/csrc/host/determinize.h"
#include "k2/csrc/host/determinize_impl.h"
#include "k2/csrc/host/fsa.h"
#include "k2/csrc/host/intersect.h"
#include "k2/csrc/host/rmepsilon.h"
#include "k2/csrc/host/topsort.h"
#include "k2/csrc/host/weights.h"
#include "k2/python/host/csrc/array.h"

void PyBindArcSort(py::module &m) {
  using PyClass = k2host::ArcSorter;
  py::class_<PyClass>(m, "_ArcSorter")
      .def(py::init<const k2host::Fsa &>(), py::arg("fsa_in"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array1<int32_t *> *arc_map = nullptr) {
            return self.GetOutput(fsa_out,
                                  arc_map == nullptr ? nullptr : arc_map->data);
          },
          py::arg("fsa_out"), py::arg("arc_map").none(true));

  m.def(
      "_arc_sort",
      [](k2host::Fsa *fsa, k2host::Array1<int32_t *> *arc_map = nullptr) {
        return k2host::ArcSort(fsa,
                               arc_map == nullptr ? nullptr : arc_map->data);
      },
      "in-place version of ArcSorter", py::arg("fsa"),
      py::arg("arc_map").none(true));
}

void PyBindTopSort(py::module &m) {
  using PyClass = k2host::TopSorter;
  py::class_<PyClass>(m, "_TopSorter")
      .def(py::init<const k2host::Fsa &>(), py::arg("fsa_in"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array1<int32_t *> *arc_map = nullptr) -> bool {
            return self.GetOutput(fsa_out,
                                  arc_map == nullptr ? nullptr : arc_map->data);
          },
          py::arg("fsa_out"), py::arg("arc_map").none(true));
}

void PyBindConnect(py::module &m) {
  using PyClass = k2host::Connection;
  py::class_<PyClass>(m, "_Connection")
      .def(py::init<const k2host::Fsa &>(), py::arg("fsa_in"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array1<int32_t *> *arc_map = nullptr) -> bool {
            return self.GetOutput(fsa_out,
                                  arc_map == nullptr ? nullptr : arc_map->data);
          },
          py::arg("fsa_out"), py::arg("arc_map").none(true));
}

void PyBindIntersect(py::module &m) {
  using PyClass = k2host::Intersection;
  py::class_<PyClass>(m, "_Intersection")
      .def(py::init<const k2host::Fsa &, const k2host::Fsa &>(),
           py::arg("fsa_a"), py::arg("fsa_b"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array1<int32_t *> *arc_map_a = nullptr,
             k2host::Array1<int32_t *> *arc_map_b = nullptr) -> bool {
            return self.GetOutput(
                fsa_out, arc_map_a == nullptr ? nullptr : arc_map_a->data,
                arc_map_b == nullptr ? nullptr : arc_map_b->data);
          },
          py::arg("fsa_out"), py::arg("arc_map_a").none(true),
          py::arg("arc_map_b").none(true));
}

template <typename TracebackState>
void PybindDeterminizerPrunedTpl(py::module &m, const char *name) {
  using PyClass = k2host::DeterminizerPruned<TracebackState>;
  py::class_<PyClass>(m, name)
      .def(py::init<const k2host::WfsaWithFbWeights &, float, int64_t>(),
           py::arg("fsa_in"), py::arg("beam"), py::arg("max_step"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"),
           py::arg("arc_derivs_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array2<typename TracebackState::DerivType *> *arc_derivs)
              -> float { return self.GetOutput(fsa_out, arc_derivs); },
          py::arg("fsa_out"), py::arg("arc_derivs"));
}

template <typename TracebackState>
void PybindEpsilonsRemoverPrunedTpl(py::module &m, const char *name) {
  using PyClass = k2host::EpsilonsRemoverPruned<TracebackState>;
  py::class_<PyClass>(m, name)
      .def(py::init<const k2host::WfsaWithFbWeights &, float>(),
           py::arg("fsa_in"), py::arg("beam"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"),
           py::arg("arc_derivs_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array2<typename TracebackState::DerivType *> *arc_derivs)
              -> void { return self.GetOutput(fsa_out, arc_derivs); },
          py::arg("fsa_out"), py::arg("arc_derivs"));
}

void PybindFsaAlgo(py::module &m) {
  PyBindArcSort(m);
  PyBindTopSort(m);
  PyBindConnect(m);
  PyBindIntersect(m);

  PybindDeterminizerPrunedTpl<k2host::MaxTracebackState>(
      m, "_DeterminizerPrunedMax");
  PybindDeterminizerPrunedTpl<k2host::LogSumTracebackState>(
      m, "_DeterminizerPrunedLogSum");

  PybindEpsilonsRemoverPrunedTpl<k2host::MaxTracebackState>(
      m, "_EpsilonsRemoverPrunedMax");
  PybindEpsilonsRemoverPrunedTpl<k2host::LogSumTracebackState>(
      m, "_EpsilonsRemoverPrunedLogSum");
}

File: k2/python/host/k2host/fsa_algo.py
# Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

# See ../../../LICENSE for clarification regarding multiple authors

import torch
from torch.utils.dlpack import to_dlpack

from .fsa import Fsa
from .array import IntArray1
from .array import IntArray2
from .array import FloatArray1
from .array import LogSumArcDerivs
from .weights import WfsaWithFbWeights
from _k2host import IntArray2Size
from _k2host import _ArcSorter
from _k2host import _arc_sort
from _k2host import _TopSorter
from _k2host import _Connection
from _k2host import _Intersection
from _k2host import _DeterminizerPrunedMax
from _k2host import _DeterminizerPrunedLogSum
from _k2host import _EpsilonsRemoverPrunedMax
from _k2host import _EpsilonsRemoverPrunedLogSum


class ArcSorter(_ArcSorter):

    def __init__(self, fsa_in: Fsa):
        super().__init__(fsa_in.get_base())

    def get_sizes(self, array_size: IntArray2Size) -> None:
        return super().get_sizes(array_size)

    def get_output(self, fsa_out: Fsa, arc_map: IntArray1 = None) -> None:
        return super().get_output(
            fsa_out.get_base(),
            arc_map.get_base() if arc_map is not None else None)


def arc_sort(fsa: Fsa, arc_map: IntArray1 = None) -> None:
    return _arc_sort(fsa.get_base(),
                     arc_map.get_base() if arc_map is not None else None)


class TopSorter(_TopSorter):

    def __init__(self, fsa_in: Fsa):
        super().__init__(fsa_in.get_base())

    def get_sizes(self, array_size: IntArray2Size) -> None:
        return super().get_sizes(array_size)

    def get_output(self, fsa_out: Fsa, arc_map: IntArray1 = None) -> bool:
        return super().get_output(
            fsa_out.get_base(),
            arc_map.get_base() if arc_map is not None else None)


class Connection(_Connection):

    def __init__(self, fsa_in: Fsa):
        super().__init__(fsa_in.get_base())

    def get_sizes(self, array_size: IntArray2Size) -> None:
        return super().get_sizes(array_size)

    def get_output(self, fsa_out: Fsa, arc_map: IntArray1 = None) -> bool:
        return super().get_output(
            fsa_out.get_base(),
            arc_map.get_base() if arc_map is not None else None)


class Intersection(_Intersection):

    def __init__(self, fsa_a: Fsa, fsa_b: Fsa):
        super().__init__(fsa_a.get_base(), fsa_b.get_base())

    def get_sizes(self, array_size: IntArray2Size) -> None:
        return super().get_sizes(array_size)

    def get_output(self,
                   fsa_out: Fsa,
                   arc_map_a: IntArray1 = None,
                   arc_map_b: IntArray1 = None) -> bool:
        return super().get_output(
            fsa_out.get_base(),
            arc_map_a.get_base() if arc_map_a is not None else None,
            arc_map_b.get_base() if arc_map_b is not None else None)


class DeterminizerPrunedMax(_DeterminizerPrunedMax):

    def __init__(self, fsa_in: WfsaWithFbWeights, beam: float, max_step: int):
        super().__init__(fsa_in, beam, max_step)

    def get_sizes(self, fsa_size: IntArray2Size,
                  arc_derivs_size: IntArray2Size) -> None:
        return super().get_sizes(fsa_size, arc_derivs_size)

    def get_output(self, fsa_out: Fsa, arc_derivs: IntArray2) -> float:
        return super().get_output(fsa_out.get_base(), arc_derivs.get_base())


class DeterminizerPrunedLogSum(_DeterminizerPrunedLogSum):

    def __init__(self, fsa_in: WfsaWithFbWeights, beam: float, max_step: int):
        super().__init__(fsa_in, beam, max_step)

    def get_sizes(self, fsa_size: IntArray2Size,
                  arc_derivs_size: IntArray2Size) -> None:
        return super().get_sizes(fsa_size, arc_derivs_size)

    def get_output(self, fsa_out: Fsa, arc_derivs: LogSumArcDerivs) -> float:
        return super().get_output(fsa_out.get_base(), arc_derivs.get_base())


class EpsilonsRemoverPrunedMax(_EpsilonsRemoverPrunedMax):

    def __init__(self, fsa_in: WfsaWithFbWeights, beam: float):
        super().__init__(fsa_in, beam)

    def get_sizes(self, fsa_size: IntArray2Size,
                  arc_derivs_size: IntArray2Size) -> None:
        return super().get_sizes(fsa_size, arc_derivs_size)

    def get_output(self, fsa_out: Fsa, arc_derivs: IntArray2) -> None:
        return super().get_output(fsa_out.get_base(), arc_derivs.get_base())


class EpsilonsRemoverPrunedLogSum(_EpsilonsRemoverPrunedLogSum):

    def __init__(self, fsa_in: WfsaWithFbWeights, beam: float):
        super().__init__(fsa_in, beam)

    def get_sizes(self, fsa_size: IntArray2Size,
                  arc_derivs_size: IntArray2Size) -> None:
        return super().get_sizes(fsa_size, arc_derivs_size)

    def get_output(self, fsa_out: Fsa, arc_derivs: LogSumArcDerivs) -> None:
        return super().get_output(fsa_out.get_base(), arc_derivs.get_base())
